### PR TITLE
Sof i915

### PR DESCRIPTION
**Pull request by libinyang**
This patch depends on #566

### CODE REVIEW

**Comment by mengdonglin** [comment]
@libinyang Would you please share more info in the body of this PR? Also please help to check and fix the CI failure. Thanks!

**Comment by libinyang** [comment]
Now that the ASoC hdmi audio PM sequence issue has been fixed by:
commit 687ae9e ("ASoC: intel: skl: Fix display power regression")
SOF need the corresponding modifications.

Meanwhile, this patch refines the error handling of display power
to release display power if probe fails.

**Comment by libinyang** [comment]
I'm not sure why Pierre's patch is in my PR. Please ignore Pierre's patch. Pierre patches have been already submitted by him. My patch is
https://github.com/thesofproject/linux/pull/569/commits/8cfc97cd7f32453780fc02aa887f6cb42fcf7362

**Comment by libinyang** [comment]
@plbossart 
Thanks for review. Please see my inline comments.

> @libinyang: not going to merge this PR as is.
> 
> 1. the first part is controversial, can you explain what is the issue you are fixing and why we no longer need to turn the display on on resume?

Before Takashi's patch "ASoC: intel: skl: Fix display power regression", hdmi audio suspend process will firstly call hdmi runtime_resume (no playback case) to turn display on and then call pci suspend to turn display off. So for resume, pci resume will turn display on and runtime_suspend will turn display off. This flow is not beautiful and sometimes it is wrong (with stress test, display power on-off pair is unbalanced which will cause suspend fail).

So we refine the hdmi audio suspend/resume flow in the patch "ASoC: intel: skl: Fix display power regression". Now the flow is very simple, display power is only handled in hdmi audio driver.

Please let me cite Takashi's comments:
" Meanwhile the ASoC hdac_hdmi driver still keeps and relies (incorrectly) on the refcounting together with ASoC skl driver"
"Basically, since the refactoring, we don't have to manage the display power at HD-audio controller suspend / resume but only at HD-audio HDMI codec suspend / resume.  So the patch drops the superfluous snd_hdac_display_power() calls in skl driver."

Correspondingly, I made a patch for sof driver like Takashi's change in skl driver.

> 2. the error flow fixes need to work even when support for HDAudio links are not compiled in. Also why would we call i915_exit which is also done in the remove()?

The error handling (hda_codec_i915_exit() and hda_deinit_caps()) is defined NULL when HDAudio links are not compiled. Please notice these function are determined by CONFIG_SND_SOC_HDAC_HDMI.
And yes, after checked the code, we needn't call i915_exit as it is done in remove(). I will remove my patch for the error handling.

**Comment by plbossart** [comment]
@libinyang thanks for the explanations. Can you update your PR so that we only use 

snd_hdac_display_power(bus, HDA_CODEC_IDX_CONTROLLER,

during
-probe (true on startup, then false at the end or error cases)
- exit (false)

all the suspend/resume cases should indeed no longer call this.

File: sound/soc/sof/intel/hda.c [hunk 1]
```diff
@@ -276,6 +276,19 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 	return 0;
 }
 
+static void hda_deinit_caps(struct snd_sof_dev *sdev)
```
Comment (ranj063) [inline sound/soc/sof/intel/hda.c]: @libinyang can we please call in hda_reset_caps() instead?

Comment (libinyang) [reply sound/soc/sof/intel/hda.c]: @ranj063 
As Pierre reminded, I will remove the hda_deinit_caps( or hda_reset_caps), as it has already been handled.

**Comment by libinyang** [comment]
> This looks good, I just have a doubt, can you double-check the flow on an error during
> 
> ```c
> ret = hda_dsp_ctrl_init_chip(sdev, true);
> 	if (ret < 0) {
> 		dev_err(bus->dev, "error: init chip failed with ret: %d\n", ret);
> 		hda_codec_i915_put(sdev); 
> 		return ret;
> 	}
> ```
> 
> in this case, we will not execute the remove() so the binding with i915 will remain. Maybe that part needs to be fixed as well.

@plbossart 
Yes, we need to handle this error. I will write another patch for this bug.

**Comment by libinyang** [comment]
> @plbossart
> Yes, we need to handle this error. I will write another patch for this bug.

https://github.com/thesofproject/linux/pull/589 has been submitted for the error handling.

**Pull request merged**